pkg/controller/volume/persistentvolume/controller_base.go

```diff
@@ -131,32 +118,46 @@ func NewPersistentVolumeController(
 	return controller
 }
 
+func NewPersistentVolumeControllerFromClient(
+	kubeClient clientset.Interface,
+	syncPeriod time.Duration,
+	provisioner vol.ProvisionableVolumePlugin,
+	recyclers []vol.VolumePlugin,
+	cloud cloudprovider.Interface,
+	clusterName string,
+	volumeSource cache.ListerWatcher,
+	claimSource cache.ListerWatcher,
+	eventRecorder record.EventRecorder,
+	enableDynamicProvisioning bool,
+) *PersistentVolumeController {
+	pvInformer := informers.NewPVInformer(kubeClient, syncPeriod)
+	if volumeSource != nil {
+		pvInformer = framework.NewSharedIndexInformer(volumeSource, &api.PersistentVolume{}, syncPeriod, cache.Indexers{"accessmodes": accessModesIndexFunc})
+	}
+	ctrl := NewPersistentVolumeController(
+		kubeClient,
+		pvInformer,
+		syncPeriod,
+		provisioner,
+		recyclers,
+		cloud,
+		clusterName,
+		claimSource,
+		eventRecorder,
+		enableDynamicProvisioning,
+	)
+	ctrl.internalPVInformer = pvInformer
+
+	return ctrl
+}
+
 // initializeCaches fills all controller caches with initial data from etcd in
 // order to have the caches already filled when first addClaim/addVolume to
 // perform initial synchronization of the controller.
-func (ctrl *PersistentVolumeController) initializeCaches(volumeSource, claimSource cache.ListerWatcher) {
-	volumeListObj, err := volumeSource.List(api.ListOptions{})
-	if err != nil {
-		glog.Errorf("PersistentVolumeController can't initialize caches: %v", err)
-		return
-	}
-	volumeList, ok := volumeListObj.(*api.PersistentVolumeList)
-	if !ok {
-		glog.Errorf("PersistentVolumeController can't initialize caches, expected list of volumes, got: %#v", volumeListObj)
-		return
-	}
-	for _, volume := range volumeList.Items {
+func (ctrl *PersistentVolumeController) initializeCaches(volumes cache.Store, claimSource cache.ListerWatcher) {
+	for _, pv := range volumes.List() {
```

Comment:
Whole point of `initializeCaches` is to fill `ctrl.volumes` with initial content. Please restore this functionality.

There are two caches, one in the (shared) informer and second in ctrl.volumes. The controller uses the later one for all its decisions and it must be initialized here. I am quite surprised it passed unit tests.